ethash/src/cache.rs

```
@@ -0,0 +1,282 @@
+	// Use uninit instead of unnecessarily writing `size_of::<Node>() * num_nodes` 0s
+	unsafe {
+		nodes.set_len(num_nodes);
+		initialize_memory(nodes.as_mut_ptr(), num_nodes, ident);
+	}
+
+	nodes
+}
+
+fn cache_path<P: ToOwned<Owned = PathBuf> + ?Sized>(path: &P, ident: &H256) -> PathBuf
+where
+	PathBuf: Borrow<P>,
+{
+	let mut buf = path.to_owned();
+	buf.push(to_hex(ident));
+	buf
+}
+
+fn consume_cache<P: AsRef<Path>>(cache: &mut Cache, path: &P) -> io::Result<()> {
+	use std::fs::OpenOptions;
+
+	let new_cache = match *cache {
+		Either::Left(ref mut vec) => {
+			let mut file = OpenOptions::new().read(true).write(true).create(true).open(&path)?;
+
+			let buf = unsafe {
+				slice::from_raw_parts_mut(vec.as_mut_ptr() as *mut u8, vec.len() * NODE_BYTES)
+			};
+
+			println!("Writing...");
+			file.write(buf).map(|_| ())?;
+
+			file
+		}
+		Either::Right(ref mmap) => {
+			println!("Flushing...");
+			mmap.flush()?;
+
+			return Ok(());
+		}
+	};
+
+	// If creating the memmap fails, we keep the `Vec` and try again next time `flush` is called.
+	// This means that it's possible to use this on a system that doesn't support memory mapping at
+	// all, at the cost of higher RAM.
+	let memmap = Mmap::open(&new_cache, Protection::ReadWrite)?;
+	*cache = Either::Right(memmap);
+
+	Ok(())
+}
+
+fn cache_from_path<P: AsRef<Path>>(path: &P) -> io::Result<Cache> {
+	Mmap::open_path(path, Protection::ReadWrite)
+		.map(Either::Right)
+		.or_else(|_| read_from_path(path).map(Either::Left))
+}
+
+fn read_from_path<P: AsRef<Path>>(path: &P) -> io::Result<Vec<Node>> {
+	use std::fs::File;
+	use std::mem;
+
+	let mut file = File::open(path)?;
+
+	let mut nodes: Vec<u8> =
+		Vec::with_capacity(file.metadata().map(|m| m.len() as _).unwrap_or(1_000_000));
+	file.read_to_end(&mut nodes)?;
+	let out: Vec<Node> = unsafe {
+		Vec::from_raw_parts(
+			nodes.as_mut_ptr() as *mut _,
+			nodes.len() / NODE_BYTES,
+			nodes.capacity() / NODE_BYTES,
```

does it actually need to return an error? just round up the capacity to nearest multiple of `NODE_BYTES` and zero-extend. doesn't matter that it'll break the calculations, but would be easier than passing an error all the way up.